File: umake.cpp
/*
 * This file is part of the umake distribution (https://github.com/nimo-labs/umake).
 * Copyright (c) 2020 Nimolabs Ltd. www.nimo.uk
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

#include "rapidjson/document.h"     // rapidjson's DOM-style API
#include "rapidjson/prettywriter.h" // for stringify JSON
#include "rapidjson/istreamwrapper.h"
#include "rapidjson/error/en.h"
#include <cstdio>
#include <string>
#include <fstream>
#include <iostream>
#include <sys/stat.h> /*mkdir */
#include <unistd.h>   /*chdir */

#include "umake.h"
#include "boilerplate.h"
#include "parseMicrocontroller.h"
#include "processTargets.h"

#define JSON_INT 0
#define JSON_STRING 1

rapidjson::Document uMakefile;
std::ofstream makefile;

using namespace rapidjson;
using namespace std;

unsigned char customLinker = 0;                       /*If 1, umakefile has defined a custom linker*/
unsigned char selectiveCheckout = CHECKOUT_TYPE_FAST; /*Default to fast checkout as selective is VEEERRRRRYYYYY slow!*/

/*Allow for overriding in-built targets*/
bool hasCustomTargetAll = 0;
bool hasCustomTargetProgram = 0;
bool hasCustomTargetReset = 0;
bool hasCustomTargetChipErase = 0;

#define PWD(a)             \
    cout << a << " pwd: "; \
    fflush(stdout);        \
    system("pwd");         \
    cout << endl           \
         << endl

void processBook(string libName, string bookName)
{
    rapidjson::Document book;
    string tmpString;

    const string strFileName = "fileName";
    const string strLanguage = "language";

    string fileName;
    string language;

    /*Get book definition*/

    if (CHECKOUT_TYPE_SELECTIVE == selectiveCheckout)
    {
        /**** Selective checkout ****/
        chdir(libName.c_str());

        string clone;
        clone.append("git checkout master -- ");
        clone.append(bookName);
        clone.append("/");

        int sysRet = system(clone.c_str());
        if (0 != sysRet) /*If we have already cloned the repo... */
            cout << "Warning: Uanable to check out book: " << bookName << endl
                 << endl;

        chdir("../");

        if (0 == bookName.compare("uC"))
            return;
        /********************/
    }

    /*Generate book descriptor filename*/
    string fn = "./umake/";
    fn.append(libName.c_str());
    fn.append("/");
    fn.append(bookName);
    fn.append("/");
    fn.append(bookName);
    fn.append(".json");

    chdir("..");

    ifstream ifs(fn);
    IStreamWrapper isw(ifs);
    book.ParseStream(isw);

    if (book.HasParseError())
    {
        fprintf(stderr, "\nError(%s: %u): %s\n",
                fn.c_str(),
                (unsigned)book.GetErrorOffset(),
                GetParseError_En(book.GetParseError()));
        exit(1);
    }

    if (!book.HasMember("book"))
    {
        cout << "Error " << fn << " is missing file field\n";
        exit(1);
    }

    /********************/
    if (!book.HasMember("book"))
    {
        cout << "Error " << fn << " is missing book field\n";
        exit(1);
    }
    tmpString = book["book"].GetString();
    if (0 != tmpString.compare(bookName))
    {
        cout << "Error " << fn << " incorrect book defined.\n";
        exit(1);
    }

    if (!book.HasMember("files"))
    {
        chdir("umake");
        return;
    }
    const rapidjson::Value &files = book["files"];
    makefile << "# " << bookName << " include path" << endl;
    makefile << "INCLUDES += -I "
             << "./umake/" << libName << "/" << bookName << "/" << endl;
    makefile << "# " << bookName << " source files" << endl;
    /*Iterate through files array*/
    for (rapidjson::Value::ConstValueIterator itr = files.Begin(); itr != files.End(); ++itr)
    {
        const rapidjson::Value &file = *itr;
        assert(file.IsObject()); // each attribute is an object
        /*Iterate through each individual file object*/
        for (rapidjson::Value::ConstMemberIterator itr2 = file.MemberBegin(); itr2 != file.MemberEnd(); ++itr2)
        {
            /*Store fileName and language*/
            if (0 == strFileName.compare(itr2->name.GetString()))
                fileName = itr2->value.GetString();
            if (0 == strLanguage.compare(itr2->name.GetString()))
                language = itr2->value.GetString();
        }

        /*output correct SRCS line to makefile*/
        if (0 == language.compare("c"))
            makefile << "SRCS += ";
        else if (0 == language.compare("c++"))
            makefile << "CPPSRCS += ";
        else
        {
            cout << "Error undefined language in book: " << bookName << endl;
            exit(1);
        }
        makefile << "./umake/" << libName << "/" << bookName << "/";
        makefile << fileName << endl;
    }

    if (book.HasMember("cflags"))
    {
        makefile << "# " << bookName << " cflags\n";
        /*Iterate through defines array*/
        const Value &cflags = book["cflags"];
        assert(cflags.IsArray());
        for (SizeType i = 0; i < cflags.Size(); i++) // Uses SizeType instead of size_t
        {
            makefile << "CFLAGS += " << cflags[i].GetString() << endl;
        }
    }

    if (book.HasMember("ldflags"))
    {
        makefile << "# " << bookName << " ldflags\n";
        assert(book.HasMember("ldflags"));
        /*Iterate through defines array*/
        const Value &ldflags = book["ldflags"];
        assert(ldflags.IsArray());
        for (SizeType i = 0; i < ldflags.Size(); i++) // Uses SizeType instead of size_t
        {
            makefile << "LDFLAGS += " << ldflags[i].GetString() << endl;
        }
    }

    chdir("umake"); /*reset our PWD so that processLibs can continue*/
}

void processLibs(void)
{
    const string strLibPath = "libPath";
    const string strLibName = "libName";

    string libName;
    string libPath;

    if (uMakefile.HasMember("libraries"))
    {
        rapidjson::Value::MemberIterator attributeIterator = uMakefile.FindMember("libraries");
        const rapidjson::Value &libraries = uMakefile["libraries"];
        assert(libraries.IsArray()); // attributes is an array

        /*Change to lib directory*/
        chdir("umake");
        /*Iterate through libraries array*/
        for (rapidjson::Value::ConstValueIterator itr = libraries.Begin(); itr != libraries.End(); ++itr)
        {
            const rapidjson::Value &library = *itr;
            assert(library.IsObject()); // each attribute is an object
            /*Iterate through each individual library object*/
            for (rapidjson::Value::ConstMemberIterator itr2 = library.MemberBegin(); itr2 != library.MemberEnd(); ++itr2)
            {
                /*Store libName and libPath*/
                if (0 == strLibName.compare(itr2->name.GetString()))
                    libName = itr2->value.GetString();
                if (0 == strLibPath.compare(itr2->name.GetString()))
                    libPath = itr2->value.GetString();
            }
            // cout << endl
            //      << "libName: " << libName << endl
            //      << "libPath: " << libPath << endl;

            string clone;

            if (CHECKOUT_TYPE_SELECTIVE == selectiveCheckout)
            {
                /***** Selective clone **********/
                clone.append("git clone --depth 1 --quiet --no-checkout --filter=blob:none ");
                clone.append(libPath);
            }
            /**************/
            else
            {
                /**** Fast clone ****/
                clone.append("git clone --depth 1 ");
                clone.append(libPath);
                /**********/
            }
            int sysRet = system(clone.c_str());
            if (0 != sysRet) /*If we have already cloned the repo... */
            {
                clone.clear();
                clone.append(libName);
                chdir(clone.c_str());
                sysRet = system("git pull");
                /*Change back*/
                chdir("..");
            }

            /*Process checkout and branch*/
            if (library.HasMember("checkout"))
            {
                chdir(libName.c_str());
                clone.clear();
                clone.append("git checkout ");
                clone.append(library["checkout"].GetString());
                cout << clone << endl;
                system(clone.c_str());
                chdir("..");
            }
            if (library.HasMember("branch"))
            {
                chdir(libName.c_str());
                clone.clear();
                clone.append("git checkout ");
                clone.append(library["branch"].GetString());
                cout << clone << endl;
                system(clone.c_str());
                chdir("..");
            }
            /*Process books for this library*/
            if (library.HasMember("books"))
            {
                makefile << "# " << libName << ": Books" << endl;
                assert(library.HasMember("books"));
                /*Iterate through defines array*/
                const Value &books = library["books"];
                assert(books.IsArray());
                for (SizeType i = 0; i < books.Size(); i++) // Uses SizeType instead of size_t
                    processBook(libName, books[i].GetString());
                makefile << endl;
            }
            if (CHECKOUT_TYPE_SELECTIVE == selectiveCheckout)
            {
                /*If we are doing selective clone then we must explicitly checkout for the uC book as well*/
                processBook(libName, "uC");
            }
        }

        /*Change back*/
        chdir("..");
    }
    else
    {
        cout << "Error. No libraries defined." << endl;
        exit(1);
    }
}

void parseUmakefile(void)
{

    if (uMakefile.HasMember("buildDir"))
    {
        makefile << "# build directory\n";
        makefile << "BUILD = " << uMakefile["buildDir"].GetString() << "\n";
    }
    else
    {
        cout << "Error. buildDir required in umakefile." << endl;
        exit(1);
    }

    if (uMakefile.HasMember("target"))
    {
        makefile << "# target binary name\n";
        makefile << "BIN = " << uMakefile["target"].GetString() << "\n\n";
    }
    else
    {
        cout << "Error. target required in umakefile." << endl;
        exit(1);
    }

    if (uMakefile.HasMember("linkerFile"))
    {
        assert(uMakefile.HasMember("linkerFile"));
        makefile << "# custom linkerfile\n";
        customLinker = 1;
        makefile << "LDFLAGS += -Wl,--script=" << uMakefile["linkerFile"].GetString() << "\n";
    }

    if (!uMakefile.HasMember("microcontroller"))
    {
        cout << "Error. microcontroller required in umakefile." << endl;
        exit(1);
    }

    if (!uMakefile.HasMember("toolchain"))
    {
        cout << "Error. toolchain required in umakefile." << endl;
        exit(1);
    }

    makefile << "# Project sources\n";
    if (uMakefile.HasMember("c_sources"))
    {
        makefile << "# C sources\n";
        assert(uMakefile.HasMember("c_sources"));
        /*Iterate through defines array*/
        const Value &sources = uMakefile["c_sources"];
        assert(sources.IsArray());
        for (SizeType i = 0; i < sources.Size(); i++) // Uses SizeType instead of size_t
        {
            makefile << "SRCS += " << sources[i].GetString() << endl;
        }
        makefile << endl;
    }

    if (uMakefile.HasMember("cpp_sources"))
    {
        makefile << "# CPP sources\n";
        assert(uMakefile.HasMember("cpp_sources"));
        /*Iterate through defines array*/
        const Value &sources = uMakefile["cpp_sources"];
        assert(sources.IsArray());
        for (SizeType i = 0; i < sources.Size(); i++) // Uses SizeType instead of size_t
        {
            makefile << "CPPSRCS += " << sources[i].GetString() << endl;
        }
        makefile << endl;
    }

    if (uMakefile.HasMember("includes"))
    {
        makefile << "# Project include directories\n";
        assert(uMakefile.HasMember("includes"));
        /*Iterate through defines array*/
        const Value &includes = uMakefile["includes"];
        assert(includes.IsArray());
        for (SizeType i = 0; i < includes.Size(); i++) // Uses SizeType instead of size_t
        {
            makefile << "INCLUDES += -I " << includes[i].GetString() << endl;
        }
        makefile << endl;
    }

    if (uMakefile.HasMember("defines"))
    {
        makefile << "# Project defines\n";
        assert(uMakefile.HasMember("defines"));
        /*Iterate through defines array*/
        const Value &includes = uMakefile["defines"];
        assert(includes.IsArray());
        for (SizeType i = 0; i < includes.Size(); i++) // Uses SizeType instead of size_t
        {
            makefile << "DEFINES += -D " << includes[i].GetString() << endl;
        }
        makefile << endl;
    }

    if (uMakefile.HasMember("cflags"))
    {
        makefile << "# Project cflags\n";
        /*Iterate through defines array*/
        const Value &cflags = uMakefile["cflags"];
        assert(cflags.IsArray());
        for (SizeType i = 0; i < cflags.Size(); i++) // Uses SizeType instead of size_t
        {
            makefile << "CFLAGS += " << cflags[i].GetString() << endl;
        }
        makefile << endl;
    }

    if (uMakefile.HasMember("ldflags"))
    {
        makefile << "# Project ldflags\n";
        assert(uMakefile.HasMember("ldflags"));
        /*Iterate through defines array*/
        const Value &ldflags = uMakefile["ldflags"];
        assert(ldflags.IsArray());
        for (SizeType i = 0; i < ldflags.Size(); i++) // Uses SizeType instead of size_t
        {
            makefile << "LDFLAGS += " << ldflags[i].GetString() << endl;
        }
        makefile << endl;
    }
}

int main(int argc, char *argv[])
{
    string line, text;

    /*Deal with umake clean*/
    if (argc > 1)
        if (0 == strncmp(argv[1], "clean", 5))
        {
            system("make clean");
            system("rm -rf ./umake");
            system("rm -f Makefile");
            exit(0);
        }

    /*Input source file*/
    ifstream in("umakefile");

    /*output Makefile*/
    makefile.open("Makefile");

    /*Create umake dir if it doesn't exist*/
    if (mkdir("umake", 0700) != 0 && errno != EEXIST)
        cout << "Error: " << errno << endl;

    /*Get umakefile*/
    ifstream ifs("umakefile");
    IStreamWrapper isw(ifs);
    uMakefile.ParseStream(isw);
    /********************/

    if (CHECKOUT_TYPE_SELECTIVE == selectiveCheckout)
        cout << "Selective checkout enabled, this may take a while!" << endl
             << endl;

    /*Process libraries*/
    processLibs();

    makefile << ".PHONY: all directory clean size\n\n";

    parseUmakefile();
    parseMicrocontroller(uMakefile, makefile, customLinker);

    /*Generate bulk of Makefile
    * Order of function calls here matters!
    */
    generateToolchainArm_none_eabi(makefile);

    processTargets(uMakefile, makefile);

    if (!hasCustomTargetAll)
        generateTargetAll(makefile);
    if (!hasCustomTargetProgram)
        generateTargetProgram(makefile);
    if (!hasCustomTargetReset)
        generateTargetReset(makefile);
    if (!hasCustomTargetChipErase)
        generateTargetChipErase(makefile);
    generateBoilerPlate(makefile);

    makefile.close();
    return 0;
}

File: umake.py
#!/usr/bin/python3

# This file is part of the hid_bootloader_console_client distribution
# (https: // github.com/nimo-labs/hid_bootloader_console_client).
# Copyright(c) 2021 Nimolabs Ltd. www.nimo.uk
#
# This program is free software: you can redistribute it and / or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see < http: // www.gnu.org/licenses/>.


import json
import os
import sys
from sys import exit
import platform
import shutil

# setting these to True inhibits generation of the corrisponding default target
inhibDefaultTgtMain = False
inhibDefaultTgtReset = False
inhibDefaultTgtChipErase = False
inhibDefaultTgtSize = False
inhibDefaultTgtClean = False
inhibDefaultTgtProgram = False
projectLinkerFile = False


def WORKING_DIR():
    return "umake"


def DEP_FILE():
    return "depfile"


def processLibs(umakefileJson, makefileHandle, depfileHandle):
    for lib in umakefileJson["libraries"]:
        currentLib = lib["libName"]
        print("Processing library: %s" % currentLib)

        if "checkout" in lib:
            cmd = "git clone --quiet  %s" % lib["libPath"]
        else:
            cmd = "git clone --quiet --depth 1 %s" % lib["libPath"]

        print(cmd)
        if 0 != os.system(cmd):
            # We've already cloned the repo, so check for updates
            os.chdir(currentLib)
            os.system("git pull")
            os.chdir("..")

        # Check to see if umakefile specifies a checkout hash
        if "checkout" in lib:
            os.chdir(currentLib)
            os.system("git checkout %s" % lib["checkout"])
            os.chdir("..")

        # Check to see if umkaefile specifies a checkout branch
        if "branch" in lib:
            os.chdir(currentLib)
            cmd = "git checkout -b %s" % lib["branch"]
            # print(cmd)
            os.system(cmd)

            # Make sure we get the latest branch commits
            # os.system("git reset --hard origin/" + lib["branch"])
            cmd = "git fetch origin " + lib["branch"]
            # print(cmd)
            os.system(cmd)

            os.system("git reset --hard FETCH_HEAD")
            os.chdir("..")

        os.chdir(currentLib)
        makefileHandle.write("# %s: books\n" % currentLib)

        for book in lib["books"]:
            currentBook = book
            os.chdir(currentBook)
            print("Processing book: %s/%s" % (currentLib, currentBook))
            bookHandle = open(currentBook+".json", 'r')
            bookJson = json.load(bookHandle)

            if bookJson['book'] != currentBook:
                print("Book: %s doesn't match current book: %s" %
                      (bookJson['book'], currentBook))
                exit()
            makefileHandle.write("# %s include path\n" % currentBook)
            makefileHandle.write(
                "INCLUDES += -I ./umake/%s/%s/\n" % (currentLib, currentBook))
            if "files" in bookJson:
                for files in bookJson['files']:
                    makefileHandle.write("# %s source files\n" % currentBook)
                    if files["language"] == 'c':
                        makefileHandle.write(
                            "SRCS += ./umake/%s/%s/%s\n" % (currentLib, currentBook, files["fileName"]))
                        depfileHandle.write(
                            "./build/%s.o: ./umake/%s/%s/%s\n" % (files["fileName"][:-2], currentLib, currentBook, files["fileName"]))
                        depfileHandle.write("\t$(UMAKE_MAKEC)\n")
                    elif files["language"] == 'cpp':
                        makefileHandle.write(
                            "CPPSRCS += ./umake/%s/%s/%s\n" % (currentLib, currentBook, files["fileName"]))
                    elif files["language"] == 'asm':
                        makefileHandle.write(
                            "ARCS += ./umake/%s/%s/%s\n" % (currentLib, currentBook, files["fileName"]))
                    else:
                        print("Unknown language for %s/%s/%s" %
                              (currentLib, currentBook, files['fileName']))
                    if "cflags" in bookJson:
                        makefileHandle.write("# Book CFLAGS\n")
                        for cflags in bookJson["cflags"]:
                            makefileHandle.write("CFLAGS += %s\n" % cflags)

                    if "ldflags" in bookJson:
                        makefileHandle.write("# Book LDFLAGS\n")
                        for ldflags in bookJson["ldflags"]:
                            makefileHandle.write(
                                "LDFLAGS += %s\n" % ldflags)

            makefileHandle.write("\n")

            bookHandle.close()
            os.chdir("..")
        os.chdir("..")


def processUc(umakefileJson, makefileHandle, depfileHandle):
    microcontroller = umakefileJson['microcontroller']
    restoreDir = os.getcwd()
    os.chdir("nimolib/uC")

    uCHandle = open("uc_"+microcontroller+"-linux-gcc.json", 'r')
    uCJson = json.load(uCHandle)

    if uCJson['microcontroller'] != microcontroller:
        print("Microcontroller: %s doesn't match requested: %s" %
              (uCJson['microcontroller'], microcontroller))
        exit()
    makefileHandle.write("# Microcontroller defines\n")
    for uCDefines in uCJson["defines"]:
        makefileHandle.write("DEFINES += -D %s\n" % uCDefines)

    makefileHandle.write("\n# Microcontroller CFLAGS\n")
    for uCCflags in uCJson["cflags"]:
        makefileHandle.write("CFLAGS += %s\n" % uCCflags)
    makefileHandle.write("\n# Microcontroller LDFLAGS\n")

    for uCLdflags in uCJson["ldflags"]:
        makefileHandle.write("LDFLAGS += %s\n" % uCLdflags)

    if False == projectLinkerFile:
        makefileHandle.write("\n# Linker file\n")
        makefileHandle.write("LDFLAGS += -Wl,--script=%s\n" %
                             uCJson['linkerFile'])


# supportFiles superseeds startupFiles
    if "supportFiles" in uCJson:
        makefileHandle.write("\n# Support files\n")
        for uCSupportFiles in uCJson["supportFiles"]:
            filename, fileExt = os.path.splitext(uCSupportFiles)
            if ".c" == fileExt:
                makefileHandle.write("SRCS += %s\n" % uCSupportFiles)
            if ".cpp" == fileExt:
                makefileHandle.write("CPPSRCS += %s\n" % uCSupportFiles)
            elif ".S" == fileExt:
                makefileHandle.write("ASRCS += %s\n" % uCSupportFiles)
            elif ".s" == fileExt:
                makefileHandle.write("ASRCS += %s\n" % uCSupportFiles)
            elif ".asm" == fileExt:
                makefileHandle.write("ASRCS += %s\n" % uCSupportFiles)

            startupFnLst = uCSupportFiles.split('/')
            startupFn = startupFnLst[len(startupFnLst)-1]
            depfileHandle.write(
                "./build/%s.o: ./%s\n" % (startupFn[:-2], uCSupportFiles))
            depfileHandle.write("\t$(UMAKE_MAKEC)\n")
    else:
        # This can probably be removed once supportFiles is implemented
        print("Warning: The startupFile entry in uC books has been deprecated, use supportFiles instead")
        makefileHandle.write("\n# Startup file\n")
        makefileHandle.write("SRCS += %s\n" % uCJson['startupFile'])
        startupFnLst = uCJson['startupFile'].split('/')
        startupFn = startupFnLst[len(startupFnLst)-1]
        depfileHandle.write(
            "./build/%s.o: ./%s\n" % (startupFn[:-2], uCJson['startupFile']))
        depfileHandle.write("\t$(UMAKE_MAKEC)\n")

    os.chdir(restoreDir)


def writeBoilerPlate(makefileHandle):
    makefileHandle.write("""CFLAGS += $(INCLUDES) $(DEFINES)
AASRCS = $(subst .s,.S, $(notdir $(ASRCS)))
ABSRCS = $(subst .asm,.S, $(notdir $(AASRCS)))

OBJS = $(addprefix $(BUILD)/, $(subst .c,.o, $(notdir $(SRCS))))
OBJS += $(addprefix $(BUILD)/, $(subst .S,.o, $(notdir $(ABSRCS))))
OBJS += $(addprefix $(BUILD)/, $(subst .cpp,.o, $(notdir $(CPPSRCS))))

define UMAKE_MAKEC
@echo CC $@
@${CC}  $(CPPFLAGS) $(CFLAGS) -c $< -o $(addprefix ${BUILD}/, $(notdir $@))
endef

define UMAKE_MAKECPP
@echo CPP $@
@${CPP}  $(CPPFLAGS) $(CFLAGS) -c $< -o $(addprefix ${BUILD}/, $(notdir $@))
endef


$(BUILD)/$(BIN).elf: $(OBJS)
	@echo LD $@
	@$(CC) $(LDFLAGS) $(addprefix ${BUILD}/, $(notdir ${OBJS})) $(LIBS) -o $@

$(BUILD)/$(BIN).hex: $(BUILD)/$(BIN).elf
	@echo OBJCOPY $@
	@$(OBJCOPY) -O ihex $^ $@

$(BUILD)/$(BIN).bin: $(BUILD)/$(BIN).elf
	@echo OBJCOPY $@
	@$(OBJCOPY) -O binary $^ $@""")


def defaultTgtMain():
    makefileHandle.write(
        "_all: $(BUILD)/$(BIN).hex $(BUILD)/$(BIN).bin size")


def defaultTgtReset():
    makefileHandle.write(
        """
reset: $(BUILD)/$(BIN).hex
	killall -s 9 openocd || true
	openocd -d1 -f ./openocd.cfg -c init -c \"reset\" -c \"exit\"""")


def defaultTgtChipErase():
    makefileHandle.write(
            """
chiperase:
	killall -s 9 openocd || true
	openocd -d1 -f ./openocd.cfg -c init -c \"at91samd chip-erase\" -c \"exit\"""")


def defaultTgtSize():
    makefileHandle.write(
        """
size: $(BUILD)/$(BIN).elf
	@echo size:
	@$(SIZE) -t $^""")


def defaultTgtClean():
    if "Linux" == platform.system():
        makefileHandle.write(
            """
clean:
\t@echo clean
\tfind ./build ! -name 'depfile' -type f -exec rm -f {} +
\t@-rm -rf ../*~""")
    else:
        makefileHandle.write(
            """
clean:
	@echo clean
	find ./build ! -name 'depfile' -type f -exec del -Force -Recurse {} +
	@-del -Force -Recurse ../*~""")


def defaultTgtProgram():
    makefileHandle.write(
        """
program: all
	hidBoot w 0x3000 $(BUILD)/$(BIN).bin""")


def remove_readonly(func, path, _):
    "Clear the readonly bit and reattempt the removal"
    os.chmod(path, stat.S_IWRITE)
    func(path)


def umakeClean(umakefileJson):
    os.system("make clean")
    if "Windows" == platform.system():
        shutil.rmtree(WORKING_DIR(), onerror=remove_readonly)
        shutil.rmtree(umakefileJson["buildDir"], onerror=remove_readonly)
        shutil.rmtree("./Makefile", onerror=remove_readonly)
    else:
        shutil.rmtree(WORKING_DIR())
        shutil.rmtree(umakefileJson["buildDir"])
        os.remove("./Makefile")


# MAIN
try:
    handle = open('./umakefile', 'r')
except:
    print("./umakefile not found.")
    exit(0)

try:
    umakefileJson = json.load(handle)
except:
    print("JSON load failed.")
    exit(0)

try:
    os.makedirs(WORKING_DIR(), exist_ok=True)
except OSError:
    print("Creation of the directory %s failed" % WORKING_DIR())

if(len(sys.argv) > 1):
    # We have command line args
    if "clean" == sys.argv[1]:
        umakeClean(umakefileJson)
        exit(0)

makefileHandle = open('Makefile', 'w')
makefileHandle.write(
    '# File autogenerated by umake, DO NOT hand edit. Changes will be overwritten\n')

# Create build directory
buildDir = umakefileJson['buildDir']
try:
    os.makedirs(buildDir, exist_ok=True)
except OSError:
    print("Creation of the directory %s failed" % buildDir)

depfileHandle = open(buildDir+"/"+DEP_FILE(), 'w')
depfileHandle.write(
    '# File autogenerated by umake, DO NOT hand edit. Changes will be overwritten\n\n')

os.chdir(WORKING_DIR())

# Process libraries
processLibs(umakefileJson, makefileHandle, depfileHandle)

# Add phony statement
makefileHandle.write("\n.PHONY: all clean size\n")
# Add build dir
makefileHandle.write("\nBUILD = %s\n" % buildDir)
# Add binary filename
makefileHandle.write("BIN = %s\n\n" % umakefileJson['target'])

# Add project level sources
makefileHandle.write("# Project sources\n")
makefileHandle.write("# C sources\n")
for projSources in umakefileJson["c_sources"]:
    rawFileName = os.path.basename(projSources)
    filename, fileExt = os.path.splitext(projSources)

    if ".c" == fileExt:
        makefileHandle.write("SRCS += %s\n" % projSources)
        rawFileName = rawFileName[:-2]
        depfileHandle.write("./build/"+rawFileName+".o: "+projSources+"\n")
        depfileHandle.write("\t$(UMAKE_MAKEC)\n")
    elif ".S" == fileExt:
        makefileHandle.write("ASRCS += %s\n" % projSources)
        rawFileName = rawFileName[:-2]
        depfileHandle.write("./build/"+rawFileName+".o: "+projSources+"\n")
        depfileHandle.write("\t$(UMAKE_MAKEC)\n")
    elif ".asm" == fileExt:
        makefileHandle.write("ASRCS += %s\n" % projSources)
        rawFileName = rawFileName[:-4]
        depfileHandle.write("./build/"+rawFileName+".o: "+projSources+"\n")
        depfileHandle.write("\t$(UMAKE_MAKEC)\n")
    elif ".cpp" == fileExt:
        makefileHandle.write("CPPSRCS += %s\n" % projSources)
        rawFileName = rawFileName[:-4]
        depfileHandle.write("./build/"+rawFileName+".o: "+projSources+"\n")
        depfileHandle.write("\t$(UMAKE_MAKECPP)\n")

makefileHandle.write("\n")


makefileHandle.write("# Project include directories\n")
for projIncludes in umakefileJson["includes"]:
    makefileHandle.write("INCLUDES += -I %s" % projIncludes)
    makefileHandle.write("\n")
makefileHandle.write("\n\n")

# Project level compiler defines
if "defines" in umakefileJson:
    makefileHandle.write("# Project level compiler defines\n")
    for defines in umakefileJson["defines"]:
        makefileHandle.write("DEFINES += %s\n" % defines)
    makefileHandle.write("\n")

# Project level LD FLAGS
if "ldflags" in umakefileJson:
    makefileHandle.write("# Project level linker flags\n")
    for ldFlags in umakefileJson["ldflags"]:
        makefileHandle.write("LDFLAGS += %s\n" % ldFlags)
    makefileHandle.write("\n")

# Custom linker file
if "linkerFile" in umakefileJson:
    projectLinkerFile = True
    makefileHandle.write("# custom linkerfile\n")
    makefileHandle.write("LDFLAGS += -Wl,--script=%s\n" %
                         umakefileJson["linkerFile"])
    makefileHandle.write("\n")

# Process microcontroller
processUc(umakefileJson, makefileHandle, depfileHandle)

# Define toolchain
makefileHandle.write("# Define toolchain\n")
makefileHandle.write("CC = " + umakefileJson["toolchain"]+"-gcc\n")
makefileHandle.write("CPP = " + umakefileJson["toolchain"]+"-g++\n")
makefileHandle.write("OBJCOPY = " + umakefileJson["toolchain"]+"-objcopy\n")
makefileHandle.write("SIZE = " + umakefileJson["toolchain"]+"-size\n")

# Primary ALL target
makefileHandle.write("\n# Primary target\n")
makefileHandle.write("all: _all\n\n")


# Custom targets
makefileHandle.write("# Custom targets\n")
if "targets" in umakefileJson:
    for custTargs in umakefileJson["targets"]:
        # Check if there is a default for this target, if so disable its generation
        if(custTargs["targetName"].lower() == "main"):
            inhibDefaultTgtMain = True
        if(custTargs["targetName"].lower() == "reset"):
            inhibDefaultTgtReset = True
        if(custTargs["targetName"].lower() == "chiperase"):
            inhibDefaultTgtChipErase = True
        if(custTargs["targetName"].lower() == "size"):
            inhibDefaultTgtSize = True
        if(custTargs["targetName"].lower() == "clean"):
            inhibDefaultTgtClean = True
        if(custTargs["targetName"].lower() == "program"):
            inhibDefaultTgtProgram = True

        # Generate custom target
        try:
            makefileHandle.write("%s: %s\n" %
                                 (custTargs["targetName"], custTargs["depends"]))
        except:
            print("Error, Target: \"%s\" missing depends key." %
                  custTargs["targetName"])
            exit(0)
        for content in custTargs["content"]:
            makefileHandle.write("\t%s\n" % content)
        makefileHandle.write("\n")


# Default targets
makefileHandle.write("# Default targets\n")
if False == inhibDefaultTgtMain:
    defaultTgtMain()
if False == inhibDefaultTgtReset:
    defaultTgtReset()
if False == inhibDefaultTgtChipErase:
    defaultTgtChipErase()
if False == inhibDefaultTgtSize:
    defaultTgtSize()
if False == inhibDefaultTgtClean:
    defaultTgtClean()
if False == inhibDefaultTgtProgram:
    defaultTgtProgram()
makefileHandle.write("\n\n")

# Boiler plate
makefileHandle.write("# Boiler plate\n")
writeBoilerPlate(makefileHandle)

# Depfile
makefileHandle.write("\n\n# Depfile\n")
makefileHandle.write("include "+buildDir+"/"+DEP_FILE()+"\n")


makefileHandle.close()
depfileHandle.close()
